### PR TITLE
Antialiasing, EPS surfaces, and PNG performance

### PR DESCRIPTION
**Pull request by asinghvi17**
This is the first stage of the refactor I've been working on.  Once this is merged into master, I can rebase the other branches I have onto this and make some manual edits.

### CODE REVIEW

File: src/CairoMakie.jl [hunk 1]
```diff
@@ -49,34 +52,48 @@ struct CairoScreen{S} <: AbstractPlotting.AbstractScreen
...
     println(io, "CairoScreen{$S} with surface:")
     println(io, screen.surface)
 end
 
-# Default to Window+Canvas as backing device
-function CairoScreen(scene::Scene)
+# Default to ARGB Surface as backing device
+# TODO: integrate Gtk into this, so we can have an interactive display
+function CairoScreen(scene::Scene; device_scaling_factor = 1, antialias = Cairo.ANTIALIAS_BEST)
     w, h = size(scene)
```
Comment (asinghvi17) [inline src/CairoMakie.jl:69]:
```suggestion
    w, h = round.(Int, scene.camera.resolution[] .* device_scaling_factor)

File: src/CairoMakie.jl [hunk 2]
```diff
@@ -817,23 +815,34 @@ function AbstractPlotting.backend_show(x::CairoBackend, io::IO, ::MIME"applicati
...
+
+function AbstractPlotting.backend_show(x::CairoBackend, io::IO, ::MIME"application/postscript", scene::Scene)
+    screen = CairoScreen(scene, io; mode=:eps)
+    cairo_draw(screen, scene)
+    Cairo.finish(screen.surface)
+    return screen
+end
+
 function AbstractPlotting.backend_show(x::CairoBackend, io::IO, m::MIME"image/png", scene::Scene)
+    screen = CairoScreen(scene; device_scaling_factor = px_per_unit)
```
Comment (asinghvi17) [inline src/CairoMakie.jl:827]:
```suggestion

**Comment by asinghvi17** [comment]
CI failures are expected and confirm that everything works correctly.

File: src/CairoMakie.jl [hunk 3]
```diff
@@ -817,23 +814,34 @@ function AbstractPlotting.backend_show(x::CairoBackend, io::IO, ::MIME"applicati
     return screen
 end
 
+
+function AbstractPlotting.backend_show(x::CairoBackend, io::IO, ::MIME"application/postscript", scene::Scene)
+    screen = CairoScreen(scene, io; mode=:eps)
```
Comment (jkrumbiegel) [inline src/CairoMakie.jl:819]: isn't the pt_per_unit thing missing here?

File: src/CairoMakie.jl [hunk 4]
```diff
@@ -372,17 +389,19 @@ function to_cairo_image(img::AbstractMatrix{<: AbstractFloat}, attributes)
...
 
 function to_cairo_image(img::Matrix{UInt32}, attributes)
-    CairoARGBSurface([img[j, i] for i in size(img, 2):-1:1, j in 1:size(img, 1)])
+    CairoARGBSurface(
+        [
+            img[j, i]
+            for i in size(img, 2):-1:1, # account for Y-axis discrepancy in Cairo
+                j in 1:size(img, 1)
+        ]
+    )
```
Comment (jkrumbiegel) [inline src/CairoMakie.jl:398]: is this the same as `collect(img')`? Maybe document more what's going on here. Probably the column major, row major conversion?

File: src/CairoMakie.jl [hunk 5]
```diff
@@ -49,34 +52,48 @@ struct CairoScreen{S} <: AbstractPlotting.AbstractScreen
...
+
     ctx = CairoContext(surf)
-    CairoScreen(scene, surf, ctx, nothing)
+    Cairo.set_antialias(ctx, antialias)
+
+    return CairoScreen(scene, surf, ctx, nothing)
 end
 
-function CairoScreen(scene::Scene, path::Union{String, IO}; mode = :svg, device_scaling_factor = 1)
+function CairoScreen(scene::Scene, path::Union{String, IO}; mode = :svg, device_scaling_factor = 1, antialias = Cairo.ANTIALIAS_BEST)
```
Comment (jkrumbiegel) [inline src/CairoMakie.jl:85]: I'm not sure that svg needs an additional antialias keyword, that seems like a png thing

File: src/CairoMakie.jl [hunk 6]
```diff
@@ -49,34 +52,48 @@ struct CairoScreen{S} <: AbstractPlotting.AbstractScreen
...
```
Comment (jkrumbiegel) [inline src/CairoMakie.jl:84]: i would just remove the `mode = :svg` default and make it a required keyword. it's not obvious why svg would be a meaningful default and then you don't know what's happening in code that calls this

File: src/CairoMakie.jl (hunk 3 above)
Comment (asinghvi17) [reply src/CairoMakie.jl:819]: Yes, good point.  Will add that in.

File: src/CairoMakie.jl (hunk 4 above)
Comment (asinghvi17) [reply src/CairoMakie.jl:398]: This is just flipping the y-axis, it needs to be transposed for conversion to a surface.  I didn't change the code - just formatted it :D.   Will add docs.

**Comment by asinghvi17** [comment]
Just rebased on master.  I think I addressed the comments you had in the commit comments, Julius (documentation of the image rotation and making the mode a required arg in `CairoScreen(scene, io, mode)`).

File: src/CairoMakie.jl [hunk 7]
```diff
@@ -49,34 +52,48 @@ struct CairoScreen{S} <: AbstractPlotting.AbstractScreen
...
 end
 
-# Default to Window+Canvas as backing device
-function CairoScreen(scene::Scene)
-    w, h = size(scene)
-    surf = CairoARGBSurface(w, h)
+# Default to ARGB Surface as backing device
+# TODO: integrate Gtk into this, so we can have an interactive display
+function CairoScreen(scene::Scene; device_scaling_factor = 1, antialias = Cairo.ANTIALIAS_BEST)
+    w, h = round.(Int, scene.camera.resolution[] .* device_scaling_factor)
```
[the remark this answers is not in the document]
Comment (asinghvi17) [reply src/CairoMakie.jl:69]: The advantage here is that you don't need to provide an unnecessary IO stream (which doesn't make sense for PNG surfaces).

**Comment by asinghvi17** [comment]
There is one test which fails because PlotUtils defines a colormap called `:grays`, and for some random reason, `to_color` can sometimes convert colors to colormaps.  Since it's an upstream issue, there's nothing we can do here.  Will make the merge, thanks Julius!

**Pull request merged**